### ISSUE TITLE
undefined symbol: _ZN6brotli16 BrotliCompressorD1Ev ?

### ISSUE THREAD

**Issue by centminmod** [google/ngx_brotli#40]
Every come across this error ? I haven't seen this before

CentOS 7 system

```
nginx -t                                                                                                                                            
nginx: [emerg] dlopen() "/usr/local/nginx/modules/ngx_http_brotli_filter_module.so" failed (/usr/local/nginx/modules/ngx_http_brotli_filter_module.so: undefined symbol: _ZN6brotli16
BrotliCompressorD1Ev) in /usr/local/nginx/conf/dynamic-modules.conf:4                                                                                                              
nginx: configuration file /usr/local/nginx/conf/nginx.conf test failed   
```

> ./configure --with-ld-opt="-lrt -ljemalloc -lpcre -Wl,-z,relro -Wl,-rpath,/usr/local/lib" --with-cc-opt="-m64 -march=native -g -O3 -fstack-protector-strong --param=ssp-buffer-size=4 -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wno-deprecated-declarations" --sbin-path=/usr/local/sbin/nginx --conf-path=/usr/local/nginx/conf/nginx.conf --with-http_stub_status_module --with-http_secure_link_module --add-module=../nginx-module-vts --with-libatomic --with-http_gzip_static_module --add-dynamic-module=../ngx_brotli --add-dynamic-module=../ngx_pagespeed-release-1.11.33.4-beta --with-http_sub_module --with-http_addition_module --with-http_image_filter_module=dynamic --with-http_geoip_module --with-stream_geoip_module --with-stream_realip_module --with-threads --with-stream=dynamic --with-stream_ssl_module --with-http_realip_module --add-dynamic-module=../ngx-fancyindex-0.4.0 --add-module=../ngx_cache_purge-2.3 --add-module=../ngx_devel_kit-0.3.0 --add-module=../set-misc-nginx-module-0.31 --add-module=../echo-nginx-module-0.60 --add-module=../redis2-nginx-module-0.13 --add-module=../ngx_http_redis-0.3.7 --add-module=../lua-nginx-module-0.10.7 --add-module=../memc-nginx-module-0.17 --add-module=../srcache-nginx-module-0.31 --add-module=../headers-more-nginx-module-0.32 --with-pcre=../pcre-8.39 --with-pcre-jit --with-http_ssl_module --with-http_v2_module --with-openssl=../libressl-2.4.4
> checking for OS
>  + Linux 3.10.0-327.28.3.el7.x86_64 x86_64
> checking for C compiler ... found
>  + using GNU C compiler
>  + gcc version: 4.8.5 20150623 (Red Hat 4.8.5-4) (GCC) 
> checking for gcc -pipe switch ... found
> checking for --with-ld-opt="-lrt -ljemalloc -lpcre -Wl,-z,relro -Wl,-rpath,/usr/local/lib" ... found
> checking for -Wl,-E switch ... found
> checking for gcc builtin atomic operations ... found
> checking for C99 variadic macros ... found
> checking for gcc variadic macros ... found
> checking for gcc builtin 64 bit byteswap ... found
> checking for unistd.h ... found
> checking for inttypes.h ... found
> checking for limits.h ... found
> checking for sys/filio.h ... not found
> checking for sys/param.h ... found
> checking for sys/mount.h ... found
> checking for sys/statvfs.h ... found
> checking for crypt.h ... found
> checking for Linux specific features
> checking for epoll ... found
> checking for EPOLLRDHUP ... found
> checking for EPOLLEXCLUSIVE ... not found
> checking for O_PATH ... found
> checking for sendfile() ... found
> checking for sendfile64() ... found
> checking for sys/prctl.h ... found
> checking for prctl(PR_SET_DUMPABLE) ... found
> checking for sched_setaffinity() ... found
> checking for crypt_r() ... found
> checking for sys/vfs.h ... found
> checking for nobody group ... found
> checking for poll() ... found
> checking for /dev/poll ... not found
> checking for kqueue ... not found
> checking for crypt() ... not found
> checking for crypt() in libcrypt ... found
> checking for F_READAHEAD ... not found
> checking for posix_fadvise() ... found
> checking for O_DIRECT ... found
> checking for F_NOCACHE ... not found
> checking for directio() ... not found
> checking for statfs() ... found
> checking for statvfs() ... found
> checking for dlopen() ... not found
> checking for dlopen() in libdl ... found
> checking for sched_yield() ... found
> checking for SO_SETFIB ... not found
> checking for SO_REUSEPORT ... found
> checking for SO_ACCEPTFILTER ... not found
> checking for SO_BINDANY ... not found
> checking for IP_BIND_ADDRESS_NO_PORT ... not found
> checking for IP_TRANSPARENT ... found
> checking for IP_BINDANY ... not found
> checking for IP_RECVDSTADDR ... not found
> checking for IP_PKTINFO ... found
> checking for IPV6_RECVPKTINFO ... found
> checking for TCP_DEFER_ACCEPT ... found
> checking for TCP_KEEPIDLE ... found
> checking for TCP_FASTOPEN ... found
> checking for TCP_INFO ... found
> checking for accept4() ... found
> checking for eventfd() ... found
> checking for int size ... 4 bytes
> checking for long size ... 8 bytes
> checking for long long size ... 8 bytes
> checking for void * size ... 8 bytes
> checking for uint32_t ... found
> checking for uint64_t ... found
> checking for sig_atomic_t ... found
> checking for sig_atomic_t size ... 4 bytes
> checking for socklen_t ... found
> checking for in_addr_t ... found
> checking for in_port_t ... found
> checking for rlim_t ... found
> checking for uintptr_t ... uintptr_t found
> checking for system byte ordering ... little endian
> checking for size_t size ... 8 bytes
> checking for off_t size ... 8 bytes
> checking for time_t size ... 8 bytes
> checking for AF_INET6 ... found
> checking for setproctitle() ... not found
> checking for pread() ... found
> checking for pwrite() ... found
> checking for pwritev() ... found
> checking for sys_nerr ... found
> checking for localtime_r() ... found
> checking for posix_memalign() ... found
> checking for memalign() ... found
> checking for mmap(MAP_ANON|MAP_SHARED) ... found
> checking for mmap("/dev/zero", MAP_SHARED) ... found
> checking for System V shared memory ... found
> checking for POSIX semaphores ... not found
> checking for POSIX semaphores in libpthread ... found
> checking for struct msghdr.msg_control ... found
> checking for ioctl(FIONBIO) ... found
> checking for struct tm.tm_gmtoff ... found
> checking for struct dirent.d_namlen ... not found
> checking for struct dirent.d_type ... found
> checking for sysconf(_SC_NPROCESSORS_ONLN) ... found
> checking for openat(), fstatat() ... found
> checking for getaddrinfo() ... found
> configuring additional modules
> adding module in ../nginx-module-vts
>  + ngx_http_vhost_traffic_status_module was configured
> adding module in ../ngx_cache_purge-2.3
>  + ngx_http_cache_purge_module was configured
> adding module in ../ngx_devel_kit-0.3.0
>  + ngx_devel_kit was configured
> adding module in ../set-misc-nginx-module-0.31
> found ngx_devel_kit for ngx_set_misc; looks good.
>  + ngx_http_set_misc_module was configured
> adding module in ../echo-nginx-module-0.60
>  + ngx_http_echo_module was configured
> adding module in ../redis2-nginx-module-0.13
>  + ngx_http_redis2_module was configured
> adding module in ../ngx_http_redis-0.3.7
>  + ngx_http_redis_module was configured
> adding module in ../lua-nginx-module-0.10.7
> checking for LuaJIT library in /usr/local/lib and /usr/local/include/luajit-2.1 (specified by the LUAJIT_LIB and LUAJIT_INC env, with -ldl) ... found
> checking for export symbols by default (-E) ... found
> checking for export symbols by default (--export-all-symbols) ... not found
> checking for SO_PASSCRED ... found
> checking for mmap(sbrk(0)) ... found
> checking for __attribute__(constructor) ... found
> checking for malloc_trim ... found
>  + ngx_http_lua_module was configured
> adding module in ../memc-nginx-module-0.17
>  + ngx_http_memc_module was configured
> adding module in ../srcache-nginx-module-0.31
>  + ngx_http_srcache_filter_module was configured
> adding module in ../headers-more-nginx-module-0.32
>  + ngx_http_headers_more_filter_module was configured
> configuring additional dynamic modules
> adding module in ../ngx_brotli
> checking for Brotli library ... found
>  + ngx_brotli was configured
> adding module in ../ngx_pagespeed-release-1.11.33.4-beta
> mod_pagespeed_dir=../ngx_pagespeed-release-1.11.33.4-beta/psol/include
> build_from_source=false
> checking for psol ... found
> List of modules (in reverse order of applicability):  ngx_http_write_filter_module ngx_http_header_filter_module ngx_http_chunked_filter_module ngx_http_v2_filter_module ngx_http_range_header_filter_module ngx_http_gzip_filter_module ngx_http_postpone_filter_module ngx_http_ssi_filter_module ngx_http_charset_filter_module ngx_http_sub_filter_module ngx_http_addition_filter_module ngx_http_userid_filter_module ngx_http_headers_filter_module
> checking for psol-compiler-compat ... found
>  + ngx_pagespeed was configured
> adding module in ../ngx-fancyindex-0.4.0
>  + ngx_http_fancyindex_module was configured
> checking for zlib library ... found
> checking for GD library ... found
> checking for GD WebP support ... not found
> checking for GeoIP library ... found
> checking for GeoIP IPv6 support ... found
> checking for atomic_ops library ... found
> creating objs/Makefile
> 
> Configuration summary
>   + using threads
>   + using PCRE library: ../pcre-8.39
>   + using OpenSSL library: ../libressl-2.4.4
>   + using system zlib library
>   + using system libatomic_ops library
> 
>   nginx path prefix: "/usr/local/nginx"
>   nginx binary file: "/usr/local/sbin/nginx"
>   nginx modules path: "/usr/local/nginx/modules"
>   nginx configuration prefix: "/usr/local/nginx/conf"
>   nginx configuration file: "/usr/local/nginx/conf/nginx.conf"
>   nginx pid file: "/usr/local/nginx/logs/nginx.pid"
>   nginx error log file: "/usr/local/nginx/logs/error.log"
>   nginx http access log file: "/usr/local/nginx/logs/access.log"
>   nginx http client request body temporary files: "client_body_temp"
>   nginx http proxy temporary files: "proxy_temp"
>   nginx http fastcgi temporary files: "fastcgi_temp"
>   nginx http uwsgi temporary files: "uwsgi_temp"
>   nginx http scgi temporary files: "scgi_temp"

**Comment by PiotrSikora** [comment]
I believe this might be because `libbrotli` installed in the system was upgraded after dynamic module was build. Does this sound possible?

I'm planning on fixing that by statically linking `brotli` during the build process (which would also eliminate 80% of the reported issues for this module), but I didn't get time to work on this yet.

**Comment by PiotrSikora** [comment]
Fixed (by bundling Brotli library) in bfd2885b2da4d763fed18f49216bb935223cd34b.
